torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py

```diff
@@ -0,0 +1,165 @@
+                                averaging, because the first process group should be a
+                                subset of the second process group, and no need to execute the first
+                                process proup redundantly.
+                                On the other hand, the third process group can only be triggered
+                                every 8 iterations, so it will not be triggered at the 4th iteration.
+                                If ``None``, the default process group,
+                                which is created by :func:`torch.distributed.init_process_group`
+                                will be used, and it will run averaging every iteration, which is
+                                normally equivalent to the vanilla gradient averaging just like
+                                :class:`~torch.nn.DistributedDataParallel` (DDP). (default: ``None``)
+        warmup_steps (int): The number of warm-up steps. During this stage, model averaging is skipped.
+        process_group (ProcessGroup, optional): The overall process group containing all the processes that runs model averaging.
+                                                If ``None``, the default process group, which is created
+                                                by :func:`torch.distributed.init_process_group`, will be used.
+                                                (default: ``None``)
+
+    Example::
+        >>>  from collections import OrderedDict
+        >>>  import torch
+        >>>  import torch.distributed as dist
+        >>>  from torch.distributed.algorithms.ddp_comm_hooks.post_localSGD_hook import (
+        >>>      PostLocalSGDState,
+        >>>      post_localSGD_hook,
+        >>>  )
+        >>>  import torch.distributed.algorithms.model_averaging.hierarchical_model_averager as hierarchicalSGD
+        >>>  import torch.nn as nn
+        >>>
+        >>>  dist.init_process_group("nccl", rank=rank, world_size=16)
+        >>>  torch.cuda.set_device(rank)
+        >>>  module = nn.Linear(1, 1, bias=False).to(rank)
+        >>>  model = nn.parallel.DistributedDataParallel(
+        >>>     module, device_ids=[rank], output_device=rank
+        >>>  )
+        >>>  # Register a post-localSGD communication hook.
+        >>>  # Assume that each machine has 4 GPUs, then each intra-machine subgroup has a size of 4.
+        >>>  subgroup, subgroups = dist.new_subgroups()
+        >>>  state = PostLocalSGDState(subgroup=subgroup, start_localSGD_iter=100)
+        >>>  model.register_comm_hook(state, post_localSGD_hook)
+        >>>
+        >>>  # Average parameters among each group of 8 processes every 4 iterations, and among all
+        >>>  # the 16 processes every 16 iterations.
+        >>>  averager = hierarchicalSGD.HierarchicalModelAverager(
+        >>>      period_group_size_dict=OrderedDict([(4, 8), (16, 16)]), warmup_steps=100)
```

Comment:
This is the best API on my mind already, which can 1) hide the complexity of creating subgroups, 2) does not require any order between <period, group size> pairs. Since this API is experimental and subject to change, we can update the API once we have a better and concrete proposal.